File: trunk/EventLib/tests/static/javascript/TestEventSerializer.js
// $Id$
//
// Unit test for event serialization and parsing
//

// Create test suite object
function TestEventSerializer() {
}

// Specify order of tests: 
TestEventSerializer.exposeTestFunctionNames = function() {
    return [ "TestEventSerializer1"
           , "testMakeSubscribeData1"
           , "testMakeSubscribeData2"
           , "testMakeSubscribeData3"
           , "testMakeSubscribeData4"
           , "testMakeUnsubscribeData"
           , "testMakeEnvelopeData1"
           , "testMakeEnvelopeData2"
           , "testMakeClosedownData"
           , "testMakeIdleData"
           , "testParseSubscribeData1"
           , "testParseSubscribeData2"
           , "testParseSubscribeData3"
           , "testParseSubscribeData4"
           , "testParseUnsubscribeData1"
           , "testParseUnsubscribeData2"
           , "testParseUnsubscribeData3"
           , "testParseUnsubscribeData4"
           , "testParseUnsubscribeData5"
           , "testParseEnvelopeData1"
           , "testParseEnvelopeData2"
           , "testParseEnvelopeData3"
           , "testParseEnvelopeData4"
           , "testParseMessageData1"
           , "testParseMessageData2"
           , "testParseMessageData3"
           , "testParseMessageData4"
           , "testParseMessageData5"
           , "testParseMessageData6"
           , "testParseMessageData7"
           , "testParseMessageData8"
           , "testParseMessageData9"
           , "testParseMessageData10"
           , "testParseMessageData11"
           , "testParseMessageData12"
           , "TestEventSerializer2"
           , "testXXX"
           ];
}

// Setup and teardown

TestEventSerializer.prototype.setUp = function() {
    this.ev  = makeEvent("s://auth.b/path/type/d#frag", 
                         "s://auth.b/path/source/d#frag", 
                         "--- payload ---");
    this.env = new EventEnvelope(this.ev,"R1").nextHop("R2").nextHop("R3");
}

TestEventSerializer.prototype.tearDown = function() {
    this.ev  = null;
    this.env = null;
}

// Test cases

TestEventSerializer.prototype.TestEventSerializer1 = function() {
    assertEquals("TestEventSerializer1", this.env.flatten(), [this.ev, ['R1', 'R2', 'R3']]);
}

TestEventSerializer.prototype.TestEventSerializer2 = function() {
    assertEquals("TestEventSerializer2", this.env.flatten(), [this.ev, ['R1', 'R2', 'R3']]);
}

// makeSubscribeData

TestEventSerializer.prototype.testMakeSubscribeData1 = function() {
    assertEquals("testMakeSubscribeData1", 
        makeSubscribeData(11, this.ev.getType(), this.ev.getSource()),
        [11, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]);
}

TestEventSerializer.prototype.testMakeSubscribeData2 = function() {
    assertEquals("testMakeSubscribeData2", 
        makeSubscribeData(22, null, this.ev.getSource()),
        [22, null, "s://auth.b/path/source/d#frag"]);
}

TestEventSerializer.prototype.testMakeSubscribeData3 = function() {
    assertEquals("testMakeSubscribeData3", 
        makeSubscribeData(33, this.ev.getType(), null),
        [33, "s://auth.b/path/type/d#frag", null]);
}

TestEventSerializer.prototype.testMakeSubscribeData4 = function() {
    assertEquals("testMakeSubscribeData4", 
        makeSubscribeData(44, null, null),
        [44, null, null]);
}

// makeUnsubscribeData, makeEnvelopeData

TestEventSerializer.prototype.testMakeUnsubscribeData = function() {
    assertEquals("testMakeUnsubscribeData", 
        makeUnsubscribeData(this.ev.getType(), this.ev.getSource()),
        [0, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]);
}

TestEventSerializer.prototype.testMakeEnvelopeData1 = function() {
    assertEquals("testMakeEnvelopeData1",
        makeEnvelopeData(this.env),
        '["forward", '+
        '[["R1", "R2", "R3"], '+
         '"s://auth.b/path/type/d#frag", '+
         '"s://auth.b/path/source/d#frag", '+
         '"--- payload ---"]]' );
}

TestEventSerializer.prototype.testMakeEnvelopeData2 = function() {
    this.ev2  = makeEvent("s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag", null);
    this.env2 = new EventEnvelope(this.ev2,"R1").nextHop("R2").nextHop("R3")
    assertEquals("testMakeEnvelopeData2",
        makeEnvelopeData(this.env2),
        '["forward", '+
        '[["R1", "R2", "R3"], '+
         '"s://auth.b/path/type/d#frag", '+
         '"s://auth.b/path/source/d#frag", '+
         'null]]');
}

// makeClosedownData, makeIdleData

TestEventSerializer.prototype.testMakeClosedownData = function() {
    assertEquals("testMakeClosedownData", makeClosedownData(), '["closedown", []]');
}

TestEventSerializer.prototype.testMakeIdleData = function() {
    assertEquals("testMakeIdleData", makeIdleData(), '["idle", []]');
}

// parseSubscribeData

TestEventSerializer.prototype.testParseSubscribeData1 = function() {
    var s = [11, null, null];
    var testps1 = parseSubscribeData(s);
    assertEquals("testParseSubscribeData1",
        [11, null, null],
        testps1);
}

TestEventSerializer.prototype.testParseSubscribeData2 = function() {
    var s = [22, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"];
    var testps2 = parseSubscribeData(s);
    assertEquals("testParseSubscribeData2",
        [22, 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag'],
        testps2);
}

TestEventSerializer.prototype.testParseSubscribeData3 = function() {
    var s = [33, null, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"];
    var testps3 = parseSubscribeData(s);
    assertEquals("testParseSubscribeData3", 
        testps3,
        [33, null, 's:\\/\\/auth.b\\/path\\/source\\/d#frag']);
}

TestEventSerializer.prototype.testParseSubscribeData4 = function() {
    var s = [44, 999, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"];
    var testps4 = parseSubscribeData(s);
    assertEquals("testParseSubscribeData4", testps4, null);
}

// parseUnsubscribeData

TestEventSerializer.prototype.testParseUnsubscribeData1 = function() {
    var s = [0, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"];
    var testpu1 = parseUnsubscribeData(s);
    assertEquals("testParseUnsubscribeData1",
        testpu1,
        ['s://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag']);
}

TestEventSerializer.prototype.testParseUnsubscribeData2 = function() {
    var s = [0, null, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"];
    var testpu2 = parseUnsubscribeData(s);
    assertEquals("testParseUnsubscribeData2",
        testpu2, 
        [null, 's:\\/\\/auth.b\\/path\\/source\\/d#frag']);
}

TestEventSerializer.prototype.testParseUnsubscribeData3 = function() {
    var s = [0, "s://auth.b/path/type/d#frag", null];
    var testpu3 = parseUnsubscribeData(s);
    assertEquals("testParseUnsubscribeData3", testpu3, ['s://auth.b/path/type/d#frag', null]);
}

TestEventSerializer.prototype.testParseUnsubscribeData4 = function() {
    var s = [0, 999, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"];
    var testpu4 = parseUnsubscribeData(s);
    assertEquals("testParseUnsubscribeData4", testpu4, null);
}

TestEventSerializer.prototype.testParseUnsubscribeData5 = function() {
    var s = [0, "s:\\/\\/auth.b\\/path\\/type\\/d#frag", 999];
    var testpu5 = parseUnsubscribeData(s);
    assertEquals("testParseUnsubscribeData5", testpu5, null);
}

// parseEnvelopeData

TestEventSerializer.prototype.testParseEnvelopeData1 = function() {
    var s = '["forward", [["R1","R2","R3"],"ev:typ","ev:src","payload"]]';
    var testpe1 = parseEnvelopeData(s);
    assertEquals("testParseEnvelopeData1", 
        testpe1,
        [["R1","R2","R3"], "ev:typ", "ev:src", "payload"]);
}

TestEventSerializer.prototype.testParseEnvelopeData2 = function() {
    var s = '["forward", [["R1",null,"R3"], null, "ev:src", null]]';
    var testpe2 = parseEnvelopeData(s);
    assertEquals("testParseEnvelopeData2", 
        testpe2,
        [["R1",null,"R3"], null, "ev:src", null]);
}

TestEventSerializer.prototype.testParseEnvelopeData3 = function() {
    var s = '["forward", [(<R1>,<R2>,<R3>),**,<ev:src>,"payload"]]';
    var testpe3 = parseEnvelopeData(s);
    assertEquals("testParseEnvelopeData3", testpe3, null);
}

TestEventSerializer.prototype.testParseEnvelopeData4 = function() {
    var s = '["forward", [(<R1>,<R2>,<R3>),<ev:typ>,**,"payload"]]';
    var testpe4 = parseEnvelopeData(s);
    assertEquals("testParseEnvelopeData4", testpe4, null);
}

// parseMessageData

TestEventSerializer.prototype.testParseMessageData1 = function() {
    var s = '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", '+
              '"ev:type", '+
              '[11, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]]]';
    var r = [ "forward", 
              [["R1","R2","R3"], 
              "http://id.webbrick.co.uk/events/subscribe",
              "ev:type", 
              [11, 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag']]];
    assertEquals("testParseMessageData1", parseMessageData(s), r);
}

TestEventSerializer.prototype.testParseMessageData2 = function() {
    var s = '["forward", '+
             '[[], '+
              '"s://auth.b/path/type/d#frag", '+
              '"s://auth.b/path/source/d#frag", '+
              'null]]';
    var r = [ 'forward', 
              [ [], 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag',
                null]];
    assertEquals("testParseMessageData2", parseMessageData(s), r);
}

TestEventSerializer.prototype.testParseMessageData3 = function() {
    var s = '["forward", '+
             '[["R1","R2","R3"], '+
              '"ev:typ", '+
              '"ev:src", '+
              '"payload"]]';
    var r = [ 'forward', 
              [ ['R1', 'R2', 'R3'], 'ev:typ', 'ev:src', 
                'payload']];
    assertEquals("testParseMessageData3", parseMessageData(s), r);
}

TestEventSerializer.prototype.testParseMessageData4 = function() {
    var s = '["forward", '+
             '[["R1","R2","R3"], '+
              '"ev:typ", '+
              '**, '+
              '"payload"]]';
    var r = null;
    assertEquals("testParseMessageData4", parseMessageData(s), r);
}

TestEventSerializer.prototype.testParseMessageData5 = function() {
    var s = '["forward", '+
             '[["R1","R2","R3"], '+
              '"ev:typ", '+
              'null, '+
              '"payload"]]';
    var r = [ 'forward', 
              [ ['R1', 'R2', 'R3'], 'ev:typ', null, 
                'payload']];
    assertEquals("testParseMessageData5", parseMessageData(s), r);
}

TestEventSerializer.prototype.testParseMessageData6 = function() {
    var s = '["closedown", []]';
    var r = [ "closedown", [] ];
    assertEquals("testParseMessageData6", parseMessageData(s), r);
}

TestEventSerializer.prototype.testParseMessageData7 = function() {
    var s = '["closedown", ["foobar"]]';
    var r = null;
    assertEquals("testParseMessageData7", parseMessageData(s), r);
};

TestEventSerializer.prototype.testParseMessageData8 = function() {
    var s = '["idle", []]';
    var r = [ "idle", [] ];
    assertEquals("testParseMessageData8", parseMessageData(s), r);
;}

TestEventSerializer.prototype.testParseMessageData9 = function() {
    var s = '["idle", ["foobar"]]';
    var r = null;
    assertEquals("testParseMessageData9", parseMessageData(s), r);
};

TestEventSerializer.prototype.testParseMessageData10 = function() {
    var s = '["idle"]';
    var r = null;
    assertEquals("testParseMessageData10", parseMessageData(s), r);
};

TestEventSerializer.prototype.testParseMessageData11 = function() {
    var s = '["idle", 666]';
    var r = null;
    assertEquals("testParseMessageData11", parseMessageData(s), r);
};

TestEventSerializer.prototype.testParseMessageData12 = function() {
    var s = '["foobar", []]';
    var r = null;
    assertEquals("testParseMessageData12", parseMessageData(s), r);
};

// XXX

TestEventSerializer.prototype.testXXX = function() {
    assertEquals("testXXX", 4, 2+2);
};

// Expose functions at global level of frame for JSUnit

// JSUnit page setup
function setUpPage() {
    info("setUpPage", "TestEventSerializer");
    setUpPageForJSUnit(TestEventSerializer, this);
}

// Return list of tests for JSunit
exposeTestFunctionNames = TestEventSerializer.exposeTestFunctionNames;

// End.

File: trunk/EventLib/tests/TestEventSerializer.py
# $Id$
#
# Copyright (c) 2008 WebBrick Systems Limited
# Released under the MIT licence
# See LICENCE.TXT included with these files,
# or http://www.opensource.org/licenses/mit-license.php
#
# Unit test for event serialization and parsing
# See http://pyunit.sourceforge.net/pyunit.html
#

import sys
import unittest
import re

sys.path.append("../..")

from EventLib.Event           import Event, makeEvent
from EventLib.EventEnvelope   import EventEnvelope
from EventLib.EventSerializer import makeEnvelopeData, makeClosedownData, makeIdleData
from EventLib.EventSerializer import parseEnvelopeData, parseMessageData
from EventLib                 import URI

es = re.compile(r"\\/")    # match escaped slash in string

# Test class
class TestEventSerializer(unittest.TestCase):

    def setUp(self):
        self.ev  = makeEvent("s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag", "--- payload ---")
        self.env = EventEnvelope(self.ev,"R1").nextHop("R2").nextHop("R3")
        return

    def tearDown(self):
        self.ev  = None
        self.env = None
        return

    # Test cases
    # NOTE: simplejson escapes forward slashes to prevent </script> attacks.

    def testEventEnvelope1(self):
        self.assertEqual( self.env.flatten(), (self.ev, ['R1', 'R2', 'R3']) )

    def testEventEnvelope2(self):
        self.assertEqual( self.env.flatten(), (self.ev, ['R1', 'R2', 'R3']) )

    def testMakeEnvelopeData1(self):
        self.assertEqual( es.sub("/", makeEnvelopeData(self.env)),
            '["forward", '+
            '[["R1", "R2", "R3"], '+
             '"s://auth.b/path/type/d#frag", '+
             '"s://auth.b/path/source/d#frag", '+
             '"--- payload ---"]]' )

    def testMakeEnvelopeData2(self):
        self.ev2  = makeEvent("s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag", None)
        self.env2 = EventEnvelope(self.ev2,"R1").nextHop("R2").nextHop("R3")
        self.assertEqual( es.sub("/", makeEnvelopeData(self.env2)),
            '["forward", '+
            '[["R1", "R2", "R3"], '+
             '"s://auth.b/path/type/d#frag", '+
             '"s://auth.b/path/source/d#frag", '+
             'null]]' )

    def testMakeEnvSubscribeData1(self):
        payload = [11, self.ev.getType(), self.ev.getSource()]
        ev      = makeEvent(URI.EventSubscribeType, "agent", payload)
        env     = EventEnvelope(ev, "router")
        self.assertEqual( es.sub("/", makeEnvelopeData(env)),
            '["forward", '+
            '[["router"], '+
             '"http://id.webbrick.co.uk/events/subscribe", '+
             '"agent", '+
             '[11, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]]]' )

    def testMakeEnvSubscribeData2(self):
        payload = [22, None, self.ev.getSource()]
        ev      = makeEvent(URI.EventSubscribeType, "agent", payload)
        env     = EventEnvelope(ev, "router")
        self.assertEqual( es.sub("/", makeEnvelopeData(env)),
            '["forward", '+
            '[["router"], '+
             '"http://id.webbrick.co.uk/events/subscribe", '+
             '"agent", '+
             '[22, null, "s://auth.b/path/source/d#frag"]]]' )

    def testMakeEnvSubscribeData3(self):
        payload = [33, self.ev.getType(), None]
        ev      = makeEvent(URI.EventSubscribeType, "agent", payload)
        env     = EventEnvelope(ev, "router")
        self.assertEqual( es.sub("/", makeEnvelopeData(env)),
            '["forward", '+
            '[["router"], '+
             '"http://id.webbrick.co.uk/events/subscribe", '+
             '"agent", '+
             '[33, "s://auth.b/path/type/d#frag", null]]]' )

    def testMakeEnvSubscribeData4(self):
        payload = [44, None, None]
        ev      = makeEvent(URI.EventSubscribeType, "agent", payload)
        env     = EventEnvelope(ev, "router")
        self.assertEqual( es.sub("/", makeEnvelopeData(env)),
            '["forward", '+
            '[["router"], '+
             '"http://id.webbrick.co.uk/events/subscribe", '+
             '"agent", '+
             '[44, null, null]]]' )

    def testMakeEnvUnsubscribeData(self):
        payload = [0, self.ev.getType(), self.ev.getSource()]
        ev      = makeEvent(URI.EventSubscribeType, "agent", payload)
        env     = EventEnvelope(ev, "router")
        self.assertEqual( es.sub("/", makeEnvelopeData(env)),
            '["forward", '+
            '[["router"], '+
             '"http://id.webbrick.co.uk/events/subscribe", '+
             '"agent", '+
             '[0, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]]]' )

    def testMakeClosedownData(self):
        self.assertEqual( makeClosedownData(), '["closedown", []]' )

    def testMakeIdleData(self):
        self.assertEqual( makeIdleData(), '["idle", []]' )

    def testParseEnvelopeData1(self):
        testpe1 = parseEnvelopeData('["forward", [["R1","R2","R3"],"ev:typ","ev:src","payload"]]')
        self.assertEqual(testpe1, [["R1","R2","R3"], "ev:typ", "ev:src", "payload"])

    def testParseEnvelopeData2(self):
        testpe2 = parseEnvelopeData('["forward", [["R1",null,"R3"], null, "ev:src", null]]')
        self.assertEqual(testpe2, [["R1",None,"R3"], None, "ev:src", None])

    def testParseEnvelopeData3(self):
        testpe3 = parseEnvelopeData('["forward", [(<R1>,<R2>,<R3>),**,<ev:src>,"payload"]]')
        assert testpe3 is None

    def testParseEnvelopeData4(self):
        testpe4 = parseEnvelopeData('["forward", [(<R1>,<R2>,<R3>),<ev:typ>,**,"payload"]]')
        assert testpe4 is None

    def testParseEnvSubscribeData1(self):
        testps = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[11, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]]]')
        self.assertEqual(testps, 
            [["R1","R2","R3"], URI.EventSubscribeType, "agent",
             [11, 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag']] )

    def testParseEnvSubscribeData2(self):
        testps = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2"], '+
              '"http:\\/\\/id.webbrick.co.uk\\/events\\/subscribe","agent", '+
              '[22, null, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"]]]')
        self.assertEqual(testps, 
            [["R1","R2"], URI.EventSubscribeType, "agent",
             [22, None, 's://auth.b/path/source/d#frag']] )

    def testParseEnvSubscribeData3(self):
        testps = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[11, **, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"]]]')
        assert testps is None

    def testParseEnvUnsubscribeData1(self):
        testpu = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[0, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]]]')
        self.assertEqual(testpu, 
            [["R1","R2","R3"], URI.EventSubscribeType, "agent",
             [0, 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag']] )

    def testParseEnvUnsubscribeData2(self):
        testpu = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[0, null, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"]]]')
        self.assertEqual(testpu, 
            [["R1","R2","R3"], URI.EventSubscribeType, "agent",
             [0, None, 's://auth.b/path/source/d#frag']] )

    def testParseEnvUnsubscribeData3(self):
        testpu = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[0, "s://auth.b/path/type/d#frag", null]]]')
        self.assertEqual(testpu, 
            [["R1","R2","R3"], URI.EventSubscribeType, "agent",
             [0, 's://auth.b/path/type/d#frag', None]] )

    def testParseEnvUnsubscribeData4(self):
        testpu = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[0, **, "s:\\/\\/auth.b\\/path\\/source\\/d#frag"]]]')
        assert testpu is None

    def testParseEnvUnsubscribeData5(self):
        testpu = parseEnvelopeData('[0, "s:\\/\\/auth.b\\/path\\/type\\/d#frag", **]')
        testpu = parseEnvelopeData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", "agent", '+
              '[0, "s:\\/\\/auth.b\\/path\\/type\\/d#frag", **]]]')
        assert testpu is None

    def testParseMessageData1(self):
        testpm1 = parseMessageData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"http://id.webbrick.co.uk/events/subscribe", '+
              '"ev:type", '+
              '[11, "s://auth.b/path/type/d#frag", "s://auth.b/path/source/d#frag"]]]')
        self.assertEqual(testpm1, 
            [ "forward", 
              [["R1","R2","R3"], 
              "http://id.webbrick.co.uk/events/subscribe",
              "ev:type", 
              [11, 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag']]])

    def testParseMessageData2(self):
        testpm2 = parseMessageData(
            '["forward", '+
             '[[], '+
              '"s://auth.b/path/type/d#frag", '+
              '"s://auth.b/path/source/d#frag", '+
              'null]]')
        self.assertEqual(testpm2, 
            ['forward', [[], 's://auth.b/path/type/d#frag', 's://auth.b/path/source/d#frag', None]])

    def testParseMessageData3(self):
        testpm3 = parseMessageData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"ev:typ", '+
              '"ev:src", '+
              '"payload"]]')
        self.assertEqual(testpm3, ['forward', [['R1', 'R2', 'R3'], 'ev:typ', 'ev:src', 'payload']])

    def testParseMessageData4(self):
        testpm4 = parseMessageData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"ev:typ", '+
              '**, '+
              '"payload"]]')
        assert testpm4 is None

    def testParseMessageData5(self):
        testpm5 = parseMessageData(
            '["forward", '+
             '[["R1","R2","R3"], '+
              '"ev:typ", '+
              'null, '+
              '"payload"]]')
        self.assertEqual(testpm5, ['forward', [['R1', 'R2', 'R3'], 'ev:typ', None, 'payload']])

    def testParseMessageData6(self):
        testpm6 = parseMessageData('["closedown", []]')
        self.assertEqual(testpm6, ['closedown',[]])
        self.assertEqual(testpm6, ['closedown',[]])

    def testParseMessageData7(self):
        testpm7 = parseMessageData('["closedown", ["foobar"]]')
        assert testpm7 is None

    def testParseMessageData8(self):
        testpm8 = parseMessageData('["idle", []]')
        self.assertEqual(testpm8, ['idle',[]])

    def testParseMessageData9(self):
        testpm9 = parseMessageData('["idle", ["foobar"]]')
        assert testpm9 is None

    def testParseMessageData10(self):
        testpm10 = parseMessageData('["idle"]')
        assert testpm10 is None

    def testParseMessageData11(self):
        testpm11 = parseMessageData('["idle", 666]')
        assert testpm11 is None

    def testParseMessageData12(self):
        testpm12 = parseMessageData('["foobar", []]')
        assert testpm12 is None

    # Sentinel/placeholder tests

    def testUnits(self):
        assert (True)

    def testComponents(self):
        assert (True)

    def testIntegration(self):
        assert (True)

    def testPending(self):
        assert (True)

# Assemble test suite
from MiscLib import TestUtils

def getTestSuite(select="unit"):
    """
    Get test suite

    select  is one of the following:
            "unit"      return suite of unit tests only
            "component" return suite of unit and component tests
            "all"       return suite of unit, component and integration tests
            "pending"   return suite of pending tests
            name        a single named test to be run
    """
    testdict = {
        "unit": 
            [ "testUnits"
            , "testEventEnvelope1"
            , "testMakeEnvelopeData1"
            , "testMakeEnvelopeData2"
            , "testMakeEnvSubscribeData1"
            , "testMakeEnvSubscribeData2"
            , "testMakeEnvSubscribeData3"
            , "testMakeEnvSubscribeData4"
            , "testMakeEnvUnsubscribeData"
            , "testMakeClosedownData"
            , "testMakeIdleData"
            , "testParseEnvelopeData1"
            , "testParseEnvelopeData2"
            , "testParseEnvelopeData3"
            , "testParseEnvelopeData4"
            , "testParseEnvSubscribeData1"
            , "testParseEnvSubscribeData2"
            , "testParseEnvSubscribeData3"
            , "testParseEnvUnsubscribeData1"
            , "testParseEnvUnsubscribeData2"
            , "testParseEnvUnsubscribeData3"
            , "testParseEnvUnsubscribeData4"
            , "testParseEnvUnsubscribeData5"
            , "testParseMessageData1"
            , "testParseMessageData2"
            , "testParseMessageData3"
            , "testParseMessageData4"
            , "testParseMessageData5"
            , "testParseMessageData6"
            , "testParseMessageData7"
            , "testParseMessageData8"
            , "testParseMessageData9"
            , "testParseMessageData10"
            , "testParseMessageData11"
            , "testParseMessageData12"
            , "testEventEnvelope2"
            ],
        "component":
            [ "testComponents"
            ],
        "integration":
            [ "testIntegration"
            ],
        "zzpending":
            [ "testPending"
            ]
        }
    return TestUtils.getTestSuite(TestEventSerializer, testdict, select=select)

# Run unit tests directly from command line
if __name__ == "__main__":
    TestUtils.runTests("TestEventSerializer.log", getTestSuite, sys.argv)

# End.
